### ISSUE TITLE
Ability to exclude items from archive based on category or tag

### ISSUE THREAD

**Issue by cdutson** [getpelican/pelican#1771]
Hi there! I've been working on converting my blog from WordPress to Pelican 3.6.0. For the most part it's been going well, but I've hit a bit of a snag.

I want to separate my content into two areas: Normal articles, and Portfolio items.

I've set the portfolio items into their own folder (and I've set the config to look at that folder for articles). This renders them into a `/category/portfolio.html` template I've knocked together with some hacky-but-usable jinja2 template stuff.

The problem is that I don't want the portfolio items showing up in the general archives. I've attempted to make a plugin to handle this, but there doesn't seem to be a signal anywhere near the archives generating that I can use. I can strip all portfolio items out of the article listing, but that then removes them entirely. I've dug through all of the generator code, and short of hacking in my own workaround, I cannot see how else I can accomplish this.

Am I over-engineering? Is there an easier way that I'm not seeing? The only thing I can think to do is to add in a template if statement to just not render portfolio items in the archive pages. The problem with this is that if I have say, 3 portfolio items in a row, then that page will only have `x-3` results to render. That's not really ideal.

**TL;DR:** I Want to exclude a specific category (or array of categories, but I'm not picky) from the archives.html template, but not from the category.html template or by removing the articles from the global articles object. I want to do so without simply hiding the items in archives.html

Thanks for any help :)

**Comment by cdutson** [comment]
Sample plugin I made that ends up being overkill. It's the closest signal I could find, but doesn't really do what I want.
```
from pelican import signals

import logging

logging.Logger(__name__)


def scrub_articles(sender):
    '''
    Given the article generator as well as using 'EXCLUDE_CATEGORIES' in the settings,
    this will remove articles from the generator that are classed as any of the excluded categories.
    '''

    excludes = sender.settings.get('EXCLUDE_CATEGORIES')
    if excludes:
        sender.articles = [x for x in sender.articles if x.category not in excludes]


def register():
    signals.article_generator_pretaxonomy.connect(scrub_articles)
```

**Comment by avaris** [comment]
Why don't you just handle it in the template (`archives.html`)? Just skip the articles with a specific category.

**Comment by cdutson** [comment]
@avaris I did cover this in my original post, but I'll provide an example as to why this isn't a great solution.

Lets say I have the following: 
- pagination set to 5
- 5 normal articles (newest)
- 5 portfolio items (older)
- 5 more articles (oldest)

Right now, if I simply hide the portfolio items in the archive code, I will end up with 1 archive pages. 1 will have 5 articles on it, and the second page will render as an empty page, with the third having 5 articles on it again.

This is obviously a more extreme example of what would go wrong, but it get's the point across.

**Comment by avaris** [comment]
Yeah, sorry. You did mention it several times. Somehow I missed it completely.

I think your best bet is to generate a custom output in a plugin. Take out `archives` from the `DIRECT_TEMPLATES` and do something like this:

```python
from pelican import signals
from operator import attrgetter

def generate_archives(generator, writer):
    excludes = generator.settings.get('EXCLUDE_CATEGORIES', [])
    articles = [x for x in generator.articles if x.category not in excludes]
    dates = sorted(articles, key=attrgetter('date'))
    context = generator.context.copy()
    context.update({'articles': articles, 'dates':dates})
    writer.write_file(
        'archives.html', 
        generator.get_template('archives'),
        context, 
        blog=True, 
        paginated={'articles': articles, 'dates': dates},
        page_name='archives',
        relative_urls=generator.settings['RELATIVE_URLS'])

def register():
    signals.article_writer_finalized.connect(generate_archives)
```

**PS**: Not tested, might have some bugs :).

**Comment by justinmayer** [comment]
@cdutson: Did Avaris's suggestion help? Any updates on this issue?

**Comment by cdutson** [comment]
@justinmayer I didn't end up using the suggestion above, I cobbled together a solution using pages and extra meta-data. I wrote briefly about it on my website: http://wallofscribbles.com/2015/07/powered-by-pelican.html under the "oddly limited" heading.

I'd be happy to expand on my solution if you'd like. I never got a chance to look into the method posted by @avaris, but I'd love to see if someone else did.

**Comment by justinmayer** [comment]
Nice work on your site and article about Pelican. A few brief comments on your write-up...

> In fact, Pelican, as near as I can tell doesn't allow for multiple categories.

I'm not sure what brought you to that conclusion, since many Pelican users publish articles in multiple categories on a daily basis. Creating sub-folders in `content` is what creates categories.

> So if you want to link to a file in a totally different area, there doesn't seem to be a way to link from root.

Linking to content root should be supported, as indicated in the [relevant section of the docs](http://docs.getpelican.com/en/latest/content.html#articles-and-pages).

That said, there are indeed areas in need of improvement, of course.  (^_^)

**Comment by cdutson** [comment]
@justinmayer I didn't really see anything in the docs re: multiple categories (I could have missed that). I do recall having issues by using the metadata method of adding categories; it was ignoring the comma. eg

`category: somthing, something else` resulted in it creating a new category called `something, something else` instead of breaking them into multiple categories.

I could not, for the life of me get linking from root to work. I read the heck out of the docs for that and just could not get it to work.

**Comment by justinmayer** [comment]
Oh, you are referring to _multiple categories assigned to one article_, which is indeed not supported. Our philosophy has been: "That is what tags are for." Categories are for broader classification (e.g, **types** of articles, etc).

If you have a reproducible list of steps that can demonstrate that linking from content root isn't functioning properly, by all means please file an issue.

**Comment by takwas** [comment]
Hello, any fix for this yet?
I have a category "Life Log" on [my blog](http://takwas.github.io/).
I don't want posts/articles in this category showing up on the homepage.

**Comment by mitchtbaum** [comment]
Could Pelican switch entirely to Jinja logic for its queries?  It would seem to provide a unified approach to building archives, any similar or unique data grouping, and what naturally needs to occur in the theme layer anyway.

Now, Jinja's docs make a good point about separating logic from templates [0].  I would answer this by saying, keep your logic elements separate, perhaps in a `{theme}/templates/chunk` directory as individual files, and call them as includes.  Finally and critically important, templates can have multiple functions, and I believe what we fundamentally want to separate here is logic from layouts.  Right now, most themes have so many different types of code intermixed, so it may seem downright senseless to implement this approach.  I do have at least [1] working demo for now, as called by [2].

0: http://jinja.pocoo.org/docs/dev/faq/#isn-t-it-a-terrible-idea-to-put-logic-into-templates
1: https://github.com/mitchtbaum/base/blob/ff5c833735ffac08a20a4097b88fdd53a64b2bc9/templates/widgets/articles-index.html
2: https://github.com/mitchtbaum/base/blob/ff5c833735ffac08a20a4097b88fdd53a64b2bc9/templates/index.html#L12

**Comment by spitis** [comment]
For anyone still looking for a fix, I was able to use this plugin to exclude a category from the rest of my blog: https://bitbucket.org/simblestudios/pelican-standalone-categories-plugin.

**Comment by justinmayer** [comment]
@spitis: Thanks for letting everyone know that there a solution available for this use case. I am subsequently closing this issue.
